Random Rewards

Mattkittan opened the pull request:
Added the ability to reward a random item to a player during the timer, as well as get a list of reward items when given a tier and major/minor outside of a timer with $rewardtable. Command names are subject to change.

Mattkittan committed:
Random

Added the ability to reward a random item to a player during the timer, as well as get a list of reward items when given a tier and major/minor outside of a timer with $rewardtable. Command names are subject to change.

MorizSchildorfer commented on cogs/reward.py:
```diff
@@ -155,7 +156,136 @@ def convert_to_seconds(s):
+    
+    # This command is called by the rewardTable command.
+    @commands.command(aliases=['rit'])
+    async def randomReward(self,ctx, tier=None, rewardType=None, amount=None):
```
You need values for these parameters, don't make them optional.
Why did you make this a public command? It does nothing for a normal user from what I see

MorizSchildorfer commented on cogs/reward.py:
```diff
@@ -155,7 +156,136 @@ def convert_to_seconds(s):
+    
+    # This is the command that is executed by the user.
+    @commands.command(aliases=['rewardtable'])
+    async def randomRewardTable(self,ctx, tier=None, rewardType=None, size=None):
```
The only one of these that seems optional is size, you later give it a default value so just set it here

MorizSchildorfer commented on cogs/reward.py:
```diff
@@ -155,7 +156,136 @@ def convert_to_seconds(s):
+        channel = ctx.channel
+        author = ctx.author
+        
+        if tier not in ('0','1','2','3','4', '5') and tier.lower() not in [r.lower() for r in roleArray]:
```
Your current defaults will lead to an error here because tier can be None

MorizSchildorfer commented on cogs/timer.py:
```diff
@@ -2174,6 +2174,168 @@ async def permissionCheck(self, msg, author):
+        rewardCheck = randomItem
+        
+        # check if the item has already been rewarded to the players
+        if (rewardCheck[rewardConsumable_group_type] in dmChar[5][1][player_type]["Major"] or
+            rewardCheck[rewardConsumable_group_type] in dmChar[5][1][player_type]["Minor"]):
+            return None # returns none, which will cause the command to be repeated to try to get a non-duplicated reward item
+        
```
There are 11 possible T5 minors, Eternal Noodle can hand out 8 of them, by running a 15h game one could hand out all possible T5 minors and then use this command and deadlock because all options have been rewarded. In any case it basically turns into a gamble on how long this command runs and that is unacceptable. There are ways to avoid the issue, though.

MorizSchildorfer commented on cogs/timer.py:
```diff
@@ -2174,6 +2174,168 @@ async def permissionCheck(self, msg, author):
+                sameMessage = True
+            return sameMessage and ((r.emoji in alphaEmojis[:6]) or (str(r.emoji) == '❌')) and u == author
+        
+        #Puts all rewards into an array. Holdover from the out-of-timer command in reward.py
```
Why is the other command not used? I ran a comparison of everything after with https://text-compare.com/ and they are literally the same...

Mattkittan replied:
You can't import commands from another file in the cogs. Or at least, I tried to and it didn't work.

MorizSchildorfer replied:
They can, the code already does so for multiple things

MorizSchildorfer commented on cogs/timer.py:
```diff
@@ -2270,6 +2432,20 @@ async def duringTimer(self,ctx, datestart, startTime, startTimes, role, game, au
+                    if await self.permissionCheck(msg, author):
+                        rewardItem = None
+                        while rewardItem == None: # This repeats if a duplicate is found.
+                            rewardItem = await self.randomRew(ctx, msg=msg, start=startTimes,dmChar=dmChar, rewardType="Major", amount=1)
```
I thought the goal was to allow multiple?

Mattkittan replied:
For the out-of-timer command.

Mattkittan committed:
Reupdate on random reward

MorizSchildorfer commented on cogs/reward.py:
```diff
@@ -4,6 +4,95 @@
+                return None
+        
+        randomItem = sample(rewardTable, int(amount)) # Makes a list of all the randomly chosen items. Includes their entire entry, which is needed for special cases of spell scrolls and ammunition
+        rewardString = []
```
Rename to be accurate about what it is

MorizSchildorfer commented on cogs/reward.py:
````diff
@@ -4,6 +4,95 @@
+                    tReaction, tUser = await self.bot.wait_for("reaction_add", check=spellEmbedCheck, timeout=60)
+                except asyncio.TimeoutError:
+                    await charEmbedmsg.delete()
+                    await channel.send(f'Spell list selection timed out! Try again using the same command:\n```yaml\n{commandPrefix}{tier}{rewardType}{amount}```')
````
The last part of the message, is that your debugging info? If yes delete it here and below

MorizSchildorfer commented on cogs/reward.py:
```diff
@@ -4,6 +4,95 @@
+                
+                output = list(spellCollection.find({"$and": [{"Classes": {"$regex": classList, '$options': 'i' }, "Level": spellLevel}]}))
+                spellReward = sample(output, 1)[0] # results in the entire spell's entry
+                spellRewardStr = []
```
Rename to be accurate

MorizSchildorfer commented on cogs/reward.py:
```diff
@@ -261,8 +267,8 @@ async def randomRewardTable(self,ctx, tier=None, rewardType=None, size=None):
         
-        if size == None:
-            size = 1
+        #if size == None:
+            #size = 1
```
delete

MorizSchildorfer commented on cogs/reward.py:
```diff
@@ -278,7 +284,8 @@ async def randomRewardTable(self,ctx, tier=None, rewardType=None, size=None):
         
-        reward = await self.randomReward(ctx, int(tierNum), rewardType, size)
+        reward = await randomReward(self, ctx, tier=int(tierNum), rewardType=rewardType, amount=size)
+        #reward = await self.randomReward(ctx, int(tierNum), rewardType, size)
```
delete

MorizSchildorfer reviewed (changes requested):
Please add a way to reward multiple at once even during the timer

Mattkittan committed:
One more update

The hopefully last PR for the random functions.

Mattkittan committed:
Merge remote-tracking branch 'upstream/master'

Mattkittan committed:
Paginate Lists

Updated the reflavorList, nicknameList, and alignmentList commands to use paginate.

MorizSchildorfer reviewed (changes requested):
Something went wrong with your merge, you overwrote timer with your outdated version
the guild stuff got removed and lots of fixes reverted

For the reflavor stuff please add back the counting. Why did you remove it?

Mattkittan commented:
> Something went wrong with your merge, you overwrote timer with your outdated version
> the guild stuff got removed and lots of fixes reverted
> 
> For the reflavor stuff please add back the counting. Why did you remove it?

I thought the merge updated the stuff, but I guess not. How would I go about undoing those changes?

I don't understand what you mean with the reflavor stuff. I thought the point of the paginate was replacing the previous outputs into something that was better? Meaning it replaced the lines involving the counting.

MorizSchildorfer commented:
> I thought the merge updated the stuff, but I guess not. How would I go about undoing those changes?
> 
> I don't understand what you mean with the reflavor stuff. I thought the point of the paginate was replacing the previous outputs into something that was better? Meaning it replaced the lines involving the counting.

You can rewind versions using git to get back to before the merge.

Paginate is meant to make the output nicer to look at and protect against expansion issues, the contents were not supposed to change. Your change removed the sorting, the duplicate reduction and the counting

MorizSchildorfer commented on cogs/admin.py:
```diff
@@ -192,50 +192,60 @@ async def alignmentList(self, ctx):
+        classContents.append(("Class", classes, False))
+        backgroundContents.append(("Background", backgrounds, False))
+            
+        await paginate(ctx, self.bot, f"Race Reflavor List", contents=raceContents, separator="\n")
+        await paginate(ctx, self.bot, f"Class Reflavor List", contents=classContents, separator="\n")
+        await paginate(ctx, self.bot, f"Background Reflavor List", contents=backgroundContents, separator="\n")
```
These should also be one paginate or else it is going to have 3 menus after each other and you can only get to next once the previous one times out

Mattkittan commented:
> You can rewind versions using git to get back to before the merge.

I'm looking there and I see no changes regarding the guild things in timer.py for some reason, or any changes made by you either.

MorizSchildorfer commented:
> I'm looking there and I see no changes regarding the guild things in timer.py for some reason, or any changes made by you either.
Yeah, because you would not have them at that point. I think what you should do is rewind to the premerge point and create a new branch from there to experiment with and see if you can get the latest timer.py on that branch. It is weird, it really seems like it just replaced the file with yours based on the changes I am seeing

Mattkittan committed:
Stuff

Reverting?

Mattkittan committed:
Merge branch 'master' into mattmaster

Mattkittan commented:
Managed to get it on a separate branch, but I can't compare it to my changes to merge it.

Mattkittan committed:
Fixes for the random

Combined, finally, maybe...
